libopae++/include/opaec++/log.h

```diff
@@ -0,0 +1,215 @@
+
+namespace opae {
+namespace fpga {
+namespace internal {
+
+/**
+ * @brief wrapped_stream wraps the logging functionality in libopae-c.
+ * It is what is returned from a logger object when it
+ * begins a stream sequence. Its destructor is called when the sequence ends.
+ */
+class wrapped_stream
+{
+ public:
+
+ /**
+  * @brief create a wrapped_stream object with a given level
+  *
+  * @param level The log level expected by the logger in libopae-c
+  * @param new_line Used to determine whether or not a newline is
+  *                 automatically appended to the stream
+  */
+  wrapped_stream(int level, bool new_line = false);
+
+  /**
+   * @brief copy constructor of wrapped_stream that copies internal
+   *        stream data structures
+   *
+   * @param other The other wrapped_stream object to copy from
+   */
+  wrapped_stream(const wrapped_stream & other);
+
+  /**
+   * @brief assignment operator of a wrapped_stream that copies internal
+   *        stream data structures
+   *
+   * @param other The other wrapped_stream object to copy from
+   *
+   * @return A reference to `this` after copying internal data
+   */
+  wrapped_stream & operator=(const wrapped_stream & other);
+
+  /**
+   * @brief wrapped_stream destructor which indicates the end of the stream
+   * sequence
+   */
+  ~wrapped_stream();
+
+  /**
+   * @brief Templated stream operator
+   *
+   * @tparam T The type of the variable being streamed
+   * @param v The variable to stream
+   *
+   * @return A reference to `this`
+   */
+  template<typename T>
+  wrapped_stream& operator<<(const T& v)
+  {
+      sstream_ << v;
+      return *this;
+  }
+
+  /**
+   * @brief Overload of the stream operator used for stream manipulators
+   *
+   * @param manip The stream manipulator
+   *
+   * @return A reference to `this`
+   */
+  wrapped_stream& operator<<(std::ostream& (*manip)(std::ostream&));
+ private:
+  std::stringstream sstream_;
+  int level_;
+  char fmt_[4];
+};
+
+/**
+ * @brief Named logger object used to format log messages with the process
+ * id, the logger name, and the level of the log message
+ */
+class logger {
+ public:
+  /**
+   * @brief log level enumeration that roughly maps to the log levels in
+   * libopae-c
+   */
+  enum class level : int
+  {
+      none        = -1,
+      fatal       = 0,
+      exception   = 10,
+      error       = 15,
+      warn        = 20,
+      info        = 25,
+      debug       = 40
+  };
+
+  /**
+   * @brief named logger constructor
+   *
+   * @param name The name of the logger to include in log messages
+   */
+  logger(const std::string & name);
+
+  /**
+   * @brief Generic log method that uses loglevel parameter
+   *
+   * @param l The level of the log message
+   * @param str An optional string that may include more contextual
+   *            information
+   *
+   * @return A wrapped_stream object to begin the stream sequence
+   */
+  wrapped_stream log(level l, std::string str = "");
```

Comment:
Do you ever want an empty log message? Why the default of "" ?

Reply:
It's mean to provide more contextual information and will get formatted by surrounding it with brackets For example:
`[16605][ERROR][logger_name][context] This is the log message`